### PR TITLE
Cli versioning for enums in flags with enum

### PR DESCRIPTION
**Pull request by rjbou**
like `--color [always|never|auto]`

### CODE REVIEW

**Comment by rjbou** [comment]
Should we keep `OpamArg.mk_tristate_opt` as the legacy  enum with `[ 'Always | 'Never | 'Auto ]` ?

File: src/client/opamArgTools.ml [hunk 1]
```diff
@@ -289,11 +289,28 @@ let mk_vflag_all ~cli ?section ?(default=[]) flags =
...
+    let flag_validity =
+      check_cli_validity cli validity ~cond:(fun c -> c && elem <> None)
+        elem flags
+    in
+    (* then check validity of the argument *)
+    match flag_validity with
+    | `Ok (Some elem) ->
+      let validity, str, _ = List.find (fun (_,_,v) -> v = elem) states in
+      check_cli_validity cli validity (Some elem)
+        [ Printf.sprintf "%s with %s" (get_long_form flags) str ]
```
Comment (dra27) [inline src/client/opamArgTools.ml:309]: The sentence reads oddly to me (I deprecated `--shell=bash` 😂 - `opam: shell with bash was ...` is initially a bit confusing). Various options:
- Put the two names in "bold" (like `OPAMCLI` and the version later in the message)
- Display `--%s=%s`
- `the %s option for --%s` (with both parts in bold)

**Review by dra27 — COMMENTED** [review]
I think a tweak to the message displayed is worth doing.

Incidentally, the same tweak could possibly be applied to the existing message:
```
dra@thor:~/opam-2$ ./opam --cli=2.1 install ocamlclean --unlock-base
opam: unlock-base was removed in version 2.1 of the opam CLI, but version 2.1 has been requested. Use --update-invariant instead or set OPAMCLI environment variable to 2.0.
```
I think it would be better for the display to be `--unlock-base` (and coloured bold in the same way as `--update-invariant` is)

**Comment by rjbou** [comment]
Initially, there is no `--` added because the error/warning function is the one used for flags, commands, and subcommands. I'll check if there is a simple way to add it.

**Comment by dra27** [comment]
Indeed - it's correct for commands subcommands - although flags could do with it as well (e.g. the `--unlock-base` example)

File: src/client/opamArgTools.ml (hunk 1 above)
Comment (rjbou) [reply src/client/opamArgTools.ml]: Finally the third one

**Comment by rjbou** [comment]
*et hop!* updated

File: src/client/opamArgTools.ml [hunk 2]
```diff
@@ -77,16 +77,23 @@ let get_long_form flags =
       if flgth > lgth then (flgth, f) else (lgth, long))
     (0,"") flags |> snd
 
-let newer_flag_msg cli valid_since elem =
+let string_of_target = function
+   | `Flags flags ->
+     OpamConsole.colorise `bold
+       "--"^get_long_form flags
```
Comment (dra27) [inline src/client/opamArgTools.ml:83]:
```suggestion
       ("--"^get_long_form flags)
```
🙂

File: src/client/opamArgTools.ml [hunk 3]
```diff
@@ -279,21 +290,39 @@ let mk_vflag_all ~cli ?section ?(default=[]) flags =
...
+        elem (`Flags flags)
+    in
+    (* then check validity of the argument *)
+    match flag_validity with
+    | `Ok (Some elem) ->
+      let validity, str, _ = List.find (fun (_,_,v) -> v = elem) states in
+      check_cli_validity cli validity (Some elem)
+        (`Verbatim
+           (Printf.sprintf "the %s option for --%s"
+              (OpamConsole.colorise `bold str)
```
Comment (dra27) [inline src/client/opamArgTools.ml:320]:
```suggestion
           (Printf.sprintf "the %s option for %s"
              (OpamConsole.colorise `bold ("--" ^ str))
```

**Review by dra27 — COMMENTED** [review]
Tweaking what gets the bold attribute, but the changes the otherwise look lovely, thanks!

**Review by dra27 — APPROVED** [review]
LGTM!

**Pull request merged**